File: src/php/RequestContext.php
<?php

declare(strict_types=1);

namespace Anonympins\Fingerprint;

/**
 * Représente le contexte d'une requête HTTP, fournissant un accès unifié
 * aux informations nécessaires pour l'analyse de l'empreinte.
 * Cette classe est conçue pour être créée à partir d'un objet de requête
 * standard (ex: PSR-7, Symfony, Laravel).
 */
class RequestContext
{
    public string $clientIp;
    public string $path;
    /** @var array<string, string> */
    public array $headers;
    /** @var array<string, mixed> */
    public array $query;
    /** @var array<string, mixed>|object|null */
    public $body;
    /** @var array<string, string> */
    public array $cookies;
    public ?string $httpVersion;
    public int $requestTimestamp;

    /** @var ?array{type: string, name: string} */
    public ?array $graphqlOperation = null;

    /** @var ?array<string, mixed> */
    public ?array $newCookieForResponse = null;

    // Propriétés spécifiques qui peuvent être fournies par un proxy inverse
    public ?string $ja3;
    public ?string $ja4;
    public ?string $http2Fingerprint;
    public ?string $tcpFingerprint;

    /**
     * @param string $clientIp
     * @param string $path
     * @param array<string, string> $headers
     * @param array<string, mixed> $query
     * @param array<string, mixed>|object|null $body
     * @param array<string, string> $cookies
     * @param string|null $httpVersion
     * @param int|null $requestTimestamp
     */
    public function __construct(
        string $clientIp,
        string $path,
        array $headers,
        array $query,
        $body,
        array $cookies,
        ?string $httpVersion,
        ?int $requestTimestamp = null
    ) {
        $this->clientIp = $clientIp;
        $this->path = $path;
        // Normaliser les en-têtes en minuscules pour un accès cohérent
        $this->headers = array_change_key_case($headers, CASE_LOWER);
        $this->query = $query;
        $this->body = $body;
        $this->cookies = $cookies;
        $this->httpVersion = $httpVersion;
        $this->requestTimestamp = $requestTimestamp ?? (int)(microtime(true) * 1000);

        // Extraire les empreintes TLS/HTTP2/TCP si elles sont fournies par les en-têtes
        $this->ja3 = $this->headers['x-ja3-hash'] ?? null;
        $this->ja4 = $this->headers['x-ja4-hash'] ?? null;
        $this->http2Fingerprint = $this->headers['x-http2-fingerprint'] ?? null;
        $this->tcpFingerprint = $this->headers['x-tcp-fingerprint'] ?? null;
    }

    public function getHeader(string $name): ?string
    {
        return $this->headers[strtolower($name)] ?? null;
    }
}

File: src/java/src/com/anonympins/fingerprint/RequestContext.java
package com.anonympins.fingerprint;

import java.util.*;

public class RequestContext {
    public String clientIp = "";
    public String path = "";
    public Map<String, String> headers = new HashMap<>();
    public Map<String, Object> queryParams = new HashMap<>();
    public Map<String, Object> body = null;
    public Map<String, String> cookies = new HashMap<>();
    public String httpVersion = "1.1";
    public long requestTimestamp;
    public String tlsSessionId = null;
    public String ja3 = null;
    public String ja4 = null;
    public String ja3Raw = null; // Raw ClientHello for server-side JA3 calculation
    public String http2Fingerprint = null;
    public String tcpFingerprint = null;
    public String quicFingerprint = null;
    public String zkpY = null;

    public RequestContext(String clientIp, String path, Map<String, String> headers,
                          Map<String, Object> queryParams, Map<String, Object> body,
                          Map<String, String> cookies, String httpVersion) {
        this.clientIp = clientIp;
        this.path = path;
        if (headers != null) {
            for (Map.Entry<String, String> entry : headers.entrySet()) {
                this.headers.put(entry.getKey().toLowerCase(), entry.getValue());
            }
        }
        this.queryParams = queryParams != null ? queryParams : new HashMap<>();
        this.body = body;
        this.cookies = cookies != null ? cookies : new HashMap<>();
        this.httpVersion = httpVersion != null ? httpVersion : "1.1";
        this.requestTimestamp = System.currentTimeMillis();

        this.ja3 = this.headers.get("x-ja3-hash");
        this.ja4 = this.headers.get("x-ja4-hash");
        this.ja3Raw = this.headers.get("x-ja3-raw");
        this.ja3 = this.headers.get("x-ja3-hash");
        this.ja4 = this.headers.get("x-ja4-hash");
        this.http2Fingerprint = this.headers.get("x-http2-fingerprint");
        this.tcpFingerprint = this.headers.get("x-tcp-fingerprint");
        this.quicFingerprint = this.headers.get("x-quic-fp");
        this.tlsSessionId = this.headers.getOrDefault("x-tls-session-id", this.headers.get("x-ssl-session-id"));
        String zkpProof = this.headers.get("x-zkp-proof");
        if (zkpProof != null && !zkpProof.isEmpty()) {
            this.zkpY = zkpProof.split(":")[0];
        }
    }

    public String getHeader(String name) {
        return this.headers.get(name.toLowerCase());
    }
}